File: 0401-0500/0401.binary-watch.cpp
#include "common.h"

class Solution {
public:
    vector<string> readBinaryWatch(int num) {
        vector<int> b1(60); b1[0] = 0;
        for (int i = 1; i < 60; ++i) {
            b1[i] = b1[i >> 1] + (i & 1);
        }
        vector<vector<int> > cs(7);
        for (int i = 0; i < 60; ++i) {
            cs[b1[i]].push_back(i);
        }
        vector<string> ans;
        // 要排除不可能的数字
        for (int u = 0; u <= min(4, num); ++u) {
            int d = num - u;
            if (d > 6) continue;
            for (int up_v : cs[u]) {
                if (up_v < 12) {
                    for (int down_v : cs[d]) {
                        if (down_v < 60) {
                            ans.push_back(to_string(up_v) + ':' + (down_v < 10 ? '0' + to_string(down_v) : to_string(down_v)));
                        } else break;
                    }
                } else break;
            }
        }
        return ans;
    }
};

int main() {
  int num; cin >> num;
  PRINT_ARRAY(Solution().readBinaryWatch(num));
  return 0;
}

File: 0401-0500/0401-binary-watch.py
import itertools

class Solution(object):
    def readBinaryWatch(self, turnedOn):
        """
        :type turnedOn: int
        :rtype: List[str]
        """
        '''
        up: 4
        down: 3
        '''
        bs = [[] for _ in range(7)]
        bs[0].append(0)
        num1s = [0]
        for i in range(1, 2 ** 6):
            d = num1s[i >> 1] + (i & 1)
            num1s.append(d)
            bs[d].append(i)
        ans = []
        for left in range(min(turnedOn, 4) + 1):
            right = turnedOn - left
            if right <= 6:
                for h, m in itertools.product(filter(lambda x: x<=11, bs[left]),
                                              filter(lambda x: x<=59, bs[right])):
                    ans.append('%d:%02d' % (h, m))
        return ans
            

print(Solution().readBinaryWatch(1))
